From: Robert Mader <robert.mader@posteo.de>
Date: Mon, 7 Nov 2022 17:07:06 +0100
Subject: wayland/actor-surface: Optimize update scheduling

This code path is important for "empty" commits to ensure we schedule
frame callbacks even if previous commits didn't cause stage redraws.
There is, however, no reason to schedule updates on all stage views
instead of only those the actor is on.

(cherry picked from commit 35ecaafd6b41bdb0e465f255801ca4e7148ad638)

Forwarded: https://gitlab.gnome.org/GNOME/mutter/-/merge_requests/2671
Forwarded: https://gitlab.gnome.org/GNOME/mutter/-/merge_requests/2763
Origin: upstream, 43.3, commit:35e4f82dbec1666f00b09898c5384ae74e0d4cce
---
 src/wayland/meta-wayland-actor-surface.c | 11 ++++++++---
 1 file changed, 8 insertions(+), 3 deletions(-)

diff --git a/src/wayland/meta-wayland-actor-surface.c b/src/wayland/meta-wayland-actor-surface.c
index 362785c..2b76d94 100644
--- a/src/wayland/meta-wayland-actor-surface.c
+++ b/src/wayland/meta-wayland-actor-surface.c
@@ -304,10 +304,15 @@ meta_wayland_actor_surface_apply_state (MetaWaylandSurfaceRole  *surface_role,
       priv->actor &&
       !meta_surface_actor_is_obscured (priv->actor))
     {
-      MetaBackend *backend = meta_get_backend ();
-      ClutterActor *stage = meta_backend_get_stage (backend);
+      GList *l;
 
-      clutter_stage_schedule_update (CLUTTER_STAGE (stage));
+      for (l = clutter_actor_peek_stage_views (CLUTTER_ACTOR (priv->actor)); l;
+           l = l->next)
+        {
+          ClutterStageView *view = l->data;
+
+          clutter_stage_view_schedule_update (view);
+        }
     }
 
   meta_wayland_actor_surface_queue_frame_callbacks (actor_surface, pending);
